File: pc/audio_track.cc
/*
 *  Copyright 2011 The WebRTC project authors. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree. An additional intellectual property rights grant can be found
 *  in the file PATENTS.  All contributing project authors may
 *  be found in the AUTHORS file in the root of the source tree.
 */

#include "pc/audio_track.h"

#include <string>

#include "absl/strings/string_view.h"
#include "api/make_ref_counted.h"
#include "api/media_stream_interface.h"
#include "api/media_stream_track.h"
#include "api/scoped_refptr.h"
#include "api/sequence_checker.h"
#include "rtc_base/logging.h"

namespace webrtc {

// static
scoped_refptr<AudioTrack> AudioTrack::Create(
    absl::string_view id,
    const scoped_refptr<AudioSourceInterface>& source) {
  return make_ref_counted<AudioTrack>(id, source);
}

AudioTrack::AudioTrack(absl::string_view label,
                       const scoped_refptr<AudioSourceInterface>& source)
    : MediaStreamTrack<AudioTrackInterface>(label), audio_source_(source) {
  if (audio_source_) {
    audio_source_->RegisterObserver(this);
    OnChanged();
  }
}

AudioTrack::~AudioTrack() {
  RTC_DCHECK_RUN_ON(&signaling_thread_checker_);
  set_state(MediaStreamTrackInterface::kEnded);
  if (audio_source_)
    audio_source_->UnregisterObserver(this);
}

std::string AudioTrack::kind() const {
  return kAudioKind;
}

AudioSourceInterface* AudioTrack::GetSource() const {
  // Callable from any thread.
  return audio_source_.get();
}

AudioProcessingOptionsResult AudioTrack::SetAudioProcessingOptions(const AudioOptions &options) {
  RTC_DCHECK_RUN_ON(&signaling_thread_checker_);
  if (!audio_source_ || audio_source_->remote()) {
    return AudioProcessingOptionsResult::Rejected(AudioProcessingOptionsResultCode::kRejectedRemoteTrack,
                                                  "Audio processing options can only be set on local audio tracks");
  }
  RTC_LOG(LS_INFO) << "AudioTrack::SetAudioProcessingOptions: " << options.ToString();

  AudioOptions processing_options;
  processing_options.echo_cancellation = options.echo_cancellation;
  processing_options.echo_cancellation_mode = options.echo_cancellation_mode;
  processing_options.noise_suppression = options.noise_suppression;
  processing_options.noise_suppression_mode = options.noise_suppression_mode;
  processing_options.auto_gain_control = options.auto_gain_control;
  processing_options.auto_gain_control_mode = options.auto_gain_control_mode;
  processing_options.highpass_filter = options.highpass_filter;
  processing_options.highpass_filter_mode = options.highpass_filter_mode;

  AudioOptions updated_options = audio_source_->options();
  updated_options.SetAll(processing_options);
  audio_source_->SetOptions(updated_options);
  FireOnChanged();
  return AudioProcessingOptionsResult::Stored();
}

void AudioTrack::AddSink(AudioTrackSinkInterface* sink) {
  RTC_DCHECK_RUN_ON(&signaling_thread_checker_);
  if (audio_source_)
    audio_source_->AddSink(sink);
}

void AudioTrack::RemoveSink(AudioTrackSinkInterface* sink) {
  RTC_DCHECK_RUN_ON(&signaling_thread_checker_);
  if (audio_source_)
    audio_source_->RemoveSink(sink);
}

void AudioTrack::OnChanged() {
  RTC_DCHECK_RUN_ON(&signaling_thread_checker_);
  if (audio_source_->state() == MediaSourceInterface::kEnded) {
    set_state(kEnded);
  } else {
    set_state(kLive);
  }
}

}  // namespace webrtc

File: sdk/android/api/org/webrtc/AudioTrack.java
/*
 *  Copyright 2013 The WebRTC project authors. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree. An additional intellectual property rights grant can be found
 *  in the file PATENTS.  All contributing project authors may
 *  be found in the AUTHORS file in the root of the source tree.
 */

package org.webrtc;

import androidx.annotation.Nullable;
import java.util.IdentityHashMap;
import org.webrtc.audio.AudioProcessingOptions;
import org.webrtc.audio.AudioProcessingOptionsResult;
import org.webrtc.audio.JavaAudioDeviceModule;

/** Java wrapper for a C++ AudioTrackInterface */
public class AudioTrack extends MediaStreamTrack {
  private final IdentityHashMap<AudioTrackSink, Long> sinks = new IdentityHashMap<AudioTrackSink, Long>();
  @Nullable private final AudioProcessingPlatformPolicy audioProcessingPlatformPolicy;

  static final class AudioProcessingPlatformPolicy {
    final boolean isEchoCancellationPlatformAvailable;
    final boolean isNoiseSuppressionPlatformAvailable;

    AudioProcessingPlatformPolicy(
        boolean isEchoCancellationPlatformAvailable, boolean isNoiseSuppressionPlatformAvailable) {
      this.isEchoCancellationPlatformAvailable = isEchoCancellationPlatformAvailable;
      this.isNoiseSuppressionPlatformAvailable = isNoiseSuppressionPlatformAvailable;
    }

    static AudioProcessingPlatformPolicy fromJavaAudioDeviceModule(
        JavaAudioDeviceModule audioDeviceModule) {
      JavaAudioDeviceModule.PlatformAudioProcessingState state =
          audioDeviceModule.getPlatformAudioProcessingState();
      return new AudioProcessingPlatformPolicy(
          state.echoCancellation.isAvailable, state.noiseSuppression.isAvailable);
    }
  }

  public AudioTrack(long nativeTrack) {
    this(nativeTrack, null);
  }

  AudioTrack(long nativeTrack, @Nullable AudioProcessingPlatformPolicy audioProcessingPlatformPolicy) {
    super(nativeTrack);
    this.audioProcessingPlatformPolicy = audioProcessingPlatformPolicy;
  }

  /** Sets the volume for the underlying MediaSource. Volume is a gain value in the range
   *  0 to 10.
   */
  public void setVolume(double volume) {
    nativeSetVolume(getNativeAudioTrack(), volume);
  }

  /** Gets the volume for the underlying MediaSource. Volume is a gain value in the range
   *  0 to 10.
   */
  public double getVolume() {
    return nativeGetVolume(getNativeAudioTrack());
  }

  /**
   * Updates the audio processing options stored on this local audio track's source.
   *
   * <p>This does not restart capture or change Android AudioRecord configuration. If the track is
   * already being sent, active senders observe the track update and reapply the updated options.
   * The effective audio processing module configuration is shared by the voice engine/channel, so
   * conflicting updates from multiple local tracks are not isolated per track.
   *
   * @return result code describing whether the request was accepted or why it was rejected. {@code
   *     STORED} means the request was accepted and stored. Rejections mean the options were not
   *     stored.
   */
  public AudioProcessingOptionsResult setAudioProcessingOptions(AudioProcessingOptions options) {
    if (options == null) {
      throw new IllegalArgumentException("AudioProcessingOptions is not allowed to be null");
    }
    // Factory-created local tracks carry the Java ADM policy, which can be
    // stricter than static device support when hardware AEC/NS was disabled by
    // the ADM builder. Tracks without that context fall back to device support.
    boolean isEchoCancellationPlatformAvailable = audioProcessingPlatformPolicy == null
        ? JavaAudioDeviceModule.isBuiltInAcousticEchoCancelerSupported()
        : audioProcessingPlatformPolicy.isEchoCancellationPlatformAvailable;
    boolean isNoiseSuppressionPlatformAvailable = audioProcessingPlatformPolicy == null
        ? JavaAudioDeviceModule.isBuiltInNoiseSuppressorSupported()
        : audioProcessingPlatformPolicy.isNoiseSuppressionPlatformAvailable;
    return AudioProcessingOptionsResult.fromNativeResult(nativeSetAudioProcessingOptions(
        getNativeAudioTrack(), options.echoCancellation,
        options.noiseSuppression, options.autoGainControl, options.highPassFilter,
        isEchoCancellationPlatformAvailable, isNoiseSuppressionPlatformAvailable,
        options.echoCancellationMode.ordinal(), options.noiseSuppressionMode.ordinal(),
        options.autoGainControlMode.ordinal(), options.highPassFilterMode.ordinal()));
  }

  /**
   * Adds an AudioTrackSink to the track. This callback is only
   * called for remote audio tracks.
   * 
   * Repeated addSink calls will not add the sink multiple times.
   */
  public void addSink(AudioTrackSink sink) {
    if (sink == null) {
      throw new IllegalArgumentException("The AudioTrackSink is not allowed to be null");
    }
    if (!sinks.containsKey(sink)) {
      final long nativeSink = nativeWrapSink(sink);
      sinks.put(sink, nativeSink);
      nativeAddSink(getNativeMediaStreamTrack(), nativeSink);
    }
  }

  /**
   * Removes an AudioTrackSink from the track.
   *
   * If the AudioTrackSink was not attached to the track, this is a no-op.
   */
  public void removeSink(AudioTrackSink sink) {
    final Long nativeSink = sinks.remove(sink);
    if (nativeSink != null) {
      nativeRemoveSink(getNativeMediaStreamTrack(), nativeSink);
      nativeFreeSink(nativeSink);
    }
  }

  @Override
  public void dispose() {
    for (long nativeSink : sinks.values()) {
      nativeRemoveSink(getNativeMediaStreamTrack(), nativeSink);
      nativeFreeSink(nativeSink);
    }
    sinks.clear();
    super.dispose();
  }

  /** Returns a pointer to webrtc::AudioTrackInterface. */
  long getNativeAudioTrack() {
    return getNativeMediaStreamTrack();
  }

  private static native void nativeSetVolume(long track, double volume);
  private static native double nativeGetVolume(long track);
  private static native String nativeSetAudioProcessingOptions(long track, boolean echoCancellation,
      boolean noiseSuppression, boolean autoGainControl, boolean highPassFilter,
      boolean isEchoCancellationPlatformAvailable, boolean isNoiseSuppressionPlatformAvailable,
      int echoCancellationMode, int noiseSuppressionMode, int autoGainControlMode,
      int highPassFilterMode);
  private static native void nativeAddSink(long track, long nativeSink);
  private static native void nativeRemoveSink(long track, long nativeSink);
  private static native long nativeWrapSink(AudioTrackSink sink);
  private static native void nativeFreeSink(long sink);
}
